keps/sig-network/3015-prefer-local-traffic-policy/README.md

```diff
@@ -0,0 +1,548 @@
+  aspects of the other two policies. A node receiving a connection to
+  a `PreferLocal` service will send it to one of the service's
+  endpoints on the same node, if possible; if there are no local
+  endpoints then it will send it to a random remote endpoint, as with
+  `Cluster`. Like `Local` it depends on an upstream load balancer to
+  distribute connections evenly, but it is slightly more reliable in
+  the face of changing endpoints, at the expense of not preserving the
+  source IP (even when delivering locally).
+
+```
+<<[UNRESOLVED renaming ]>>
+
+Should we add a new alias for `Local`, such as `PreserveSourceIP`, and
+deprecate the old name?
+
+<<[/UNRESOLVED]>>
+```
+
+[public documentation on externalTrafficPolicy]: https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip
+
+### Internal Traffic Policy
+
+This is easy; source IP is always preserved for internal traffic, so
+`PreferLocal` can be implemented in the obvious way with no caveats.
+
+```
+<<[UNRESOLVED internal-traffic-policy-local ]>>
+
+Given `PreferLocal`, is there any reason to keep the behavior of
+`internalTrafficPolicy: Local` with no fallback? Is there actually a
+use case for "either send locally or drop the traffic"?
+
+The KEP gives the use case:
+
+    As a platform owner, I want to create a Service that always
+    directs traffic to a logging daemon on the same node. Traffic
+    should never bounce to a daemon on another node.
+
+But there's an implied "and if the logging daemon is unavailable, then
+logs should be dropped". Are there really cases where you want that?
+And the answer can't be "the administrator will just make sure that
+the logging daemon is always available", because (a) you can't do
+that, and (b) if you could do that then `PreferLocal` would give the
+same result as `Local` anyway, since it would never need to fall back.
+
+The strongest argument for keeping distinct `Local` and `PreferLocal`
+in the internal traffic case is that it makes it parallel with the
+external traffic case (other than the source IP stuff). But I feel
+like we really need to document "you almost never want
+`internalTrafficPolicy: Local`".
+
+If we rename `externalTrafficPolicy: Local` then there's a stronger
+argument for deprecating `internalTrafficPolicy: Local` along with it.
+
+<<[/UNRESOLVED]>>
+```
+
+### Interaction with Topology
+
+The [Topology Aware Hints KEP] says "Both ExternalTrafficPolicy and
+InternalTrafficPolicy will be given precedence over topology aware
+routing." (By which it really means that _`Local`_ traffic policy has
+precedence over topology, and topology only applies to `Cluster`
+traffic policy.)
+
+When delivering packets locally, `PreferLocal` policy will have to
+ignore topology, just like `Local` policy does.
+
+When delivering packets remotely, it makes the most sense to have
+`PreferLocal` policy make use of topology. (This is explicitly wanted
```

Comment:
One of the reasons we scrapped PreferLocal in the other KEP is because the assumption was you can achieve the same behavior by setting internalTrafficPolicy: Cluster and topology-aware routing that strongly prefers node-local endpoints. What is the semantic difference between that and using PreferLocal?

Reply:
If we had node-level topology, then we probably would not need `PreferLocal`... but do we even have a plan for node-level topology yet?

Reply:
Arguably if we had node-level topology, then we would not need `internalTrafficPolicy: Local` either, since (unlike `externalTrafficPolicy: Local`) it provides no advantages other than locality.

Reply:
I'll attach my main feedback here.

Let's consider: Why do people use [EI]TP ?

ETP:  I know of 2 reasons.  1) "I _need_ to preserve source IP".  2) "Efficiency".

ITP: I know of 2 reasons.  1) "I _need_ to stay on this node (semantic correctness)".  2) "Efficiency".

For both classes of traffic, "Local" is either a hard requirement (i.e. "prefer" is a non-starter) or just an optimization over "Cluster".

I don't like that we put the responsibility onto our users tell us which optimizations we should be using.  That seems like a total failure on our part.  OF COURSE they want us to optimize if we can.  The specific optimization that PreferLocal dictates is pretty black or white, which means there's not much room for being "smart".  This then FORCES other implementations, which might be capable of more smartness, to play dumb.

For example, it's really not hard to imagine an implementation that sends the number of active flows per-endpoint to an aggregator, and then have that aggregator give the node-agents hints about which endpoints are overloaded.  It doesn't even need to be super timely to have some impact, I bet.  But the semantics of "PreferLocal" disallow that.  

If we think this is the optimizatrion with the biggest bang for the buck, let's talk about how we can do something like `if num_endpoints > (num_nodes/2)  && have_local_eps() { use_local_eps() }`.  I just made that up - obviously we can tune the heuristic.

Reply:
For our use cases, mostly daemonsets providing platform infrastructure, we definitely have more _prefer_ local endpoints than _needs_ local endpoints. The current semantics for ITP "Local" is very bad for our users, as the daemonset pod becoming not ready provides not optimization but a production outage (in a form of dns cache, for example). Conversely, ITP "Cluster" and Topology Aware Hints also provide the wrong semantics, where we're more likely to not send traffic locally.

We've had to leave ITP as "Cluster" and have external tools try to update iptables when the daemonset pods are ready, and remove when they are not, which is a fragile situation we'd prefer not to be in. We were looking forward to PreferLocal in the previous proposal, as not having it causes us pain today.

Many of our users, who are running business workflows in deployments, prefer semantics of "Cluster" and Topology Aware Hints. Trying to be "smart" can potentially hotspot specific instances just because they are on the same node as an ingress controller or another workload.

> If we think this is the optimizatrion with the biggest bang for the buck, let's talk about how we can do something like `if num_endpoints > (num_nodes/2) && have_local_eps() { use_local_eps() }`.

I worry a heuristic along this pattern would be surprising to our users. Currently, most aspects of scaling clusters up and down are transparent operations to them. These types of heuristics would suddenly change traffic patterns for them, in potentially unexpected ways.

Reply:
@thockin just to see if I get your point, you mean that PreferLocal is a loadbalancing/topology algorithm problem and not a traffic policy problem, hence this is not the right place to fix it?

Reply:
> PreferLocal is a loadbalancing/topology algorithm problem and not a traffic policy problem, hence this is not the right place to fix it

@aojea That's a reasonable distillation :)

@terinjokes thanks for the additional input.  "Local" gets better with the terminating endpoints work that is in flight, but still there's a moment when it has to die for a successor to start.  If we really want to make node-level "sidecars" a thing that is reliable, we need a way to "hand off".

But such a mechanism does (kind of) exist - the `updateStrategy` of a DaemonSet allows `maxSurge` to be positive, so we can start a new replica before killing an old one.  This is subject to resource limitations on heavily packed nodes.  If you think that is you, you can choose to run 2 DaemonSets (e.g. "dns-1" and "dns-a") so you pre-allocate space.  Maybe DaemonSet actually needs "replicas" field to run N copies per target node...

Topology, as implemented today, is not really considering node.  It could, and yes my proposed heuristic is far from  perfect.  But not considering node was sort of intentional.  In my experience, such optimizations quietly become functional requirements.  Apps (or SLOs) start to _depend_ on the optimization, and when the optimization isn't available for some reason, apps break.  

So my intuitive lean is to either NOT optimize that way, or else to make it really clear that it is an optimization, not a policy.  If you _depend_ on the optimization, then "prefer" is wrong, and we should focus on making "Local" work for you.  If you don't depend on it, then you should let the system try to be smart, and not tell it how to do its job.

That said I really do understand how "PreferLocal" is attractive, especially while the system is NOT as smart as you want it to be.  On this topic, I have cited this before: https://en.wikipedia.org/wiki/Attractive_nuisance_doctrine .  As you said `Trying to be "smart" can potentially hotspot specific instances`.  It's an overly-sharp tool that SEEMS like a good idea, but probably isn't. 

In some sense what you are asking for is not an abstract policy, but an implementation-centric toggle that says "I know what I am doing and I know how the implementation works".  If we put that into the API we get to support it FOREVER.  We reduce the degrees of freedom implementations have for doing good things, and we lock them into sub-optimal semantics.

Reply:
> ITP: I know of 2 reasons. 1) "I need to stay on this node (semantic correctness)". 2) "Efficiency".

`internalTrafficPolicy: Local` doesn't solve the "efficiency" use case though, because it implies "I would rather lose packets than let this connection go to another node", which is never what you want for the "efficiency" use case.

> I don't like that we put the responsibility onto our users tell us which optimizations we should be using. That seems like a total failure on our part. OF COURSE they want us to optimize if we can.

But optimize for _what_?

"Cluster" traffic policy optimizes under the assumption that your service will work best if kube-proxy distributes connections evenly across all the endpoints. This might be the best choice if each connection creates a lot of load on the endpoint, so overloading a local endpoint would slow response time down by more than forwarding requests to another node would.

"PreferLocal" traffic policy optimizes under the assumption that the service will work best if kube-proxy routes connections to the closest endpoint. This is the right choice for something like DNS where connections are short lived, and processing them doesn't require much CPU, so it doesn't matter if lots of connections get sent to the same endpoint all at once.

> The specific optimization that PreferLocal dictates is pretty black or white, which means there's not much room for being "smart".

Yeah, because actually I lied above: "PreferLocal" doesn't say "route connections to the closest endpoint", it says "route connections to a local endpoint, or if you can't, then route it anywhere", which is better than "Cluster"'s behavior, but not really exactly what you want.

> This then FORCES other implementations, which might be capable of more smartness, to play dumb.

In theory an implementation might be super smart and actively monitor the cluster to figure out the best way to route connections, but I think it's more likely that you'll get good behavior if we let the user give the proxy implementation hints. We just need to figure out the right set of hints.

I guess having traffic policy encode both behavioral changes ("preserve source IP", "must stay on node for semantic correctness") _and_ efficiency hints is probably a bad approach. But in that case we should clarify that traffic policy is only describing desired behavior, not desired efficiency.

Reply:
I know of cases where falling back to a non-local endpoint is semantically wrong, and isn't acceptable.  Setting a policy to "Local" ought to be a policy, not a hint.  *IF* we support "PreferLocal", it has to be disticnt from "Local".  It still seems wrong to conflate policy and efficiency hints.  

E.g. imagine a near-ideal implementation.  It would gather stats about how many connections and bytes were sent per usint time.  Maybe it would consider latency (cracking flows to recognize DNS or maybe just based on ports).  Maybe it would consider CPU load on the server.  It would use that to make a "good" choice.  Hopefully "pods on this node" would be the choice whenever possible.  We don't need a user to tell us that.

Now, we don't have a near-ideal implementation, so should we:

a) give a policy knob that a closer-to-ideal solution might wrestle with

b) give a policy knob that other implementations can choose to ignore (e.g. annotation);

c) let the user offer a hint that is explicitly a hint (which implementations can choose to respect or not)

d) start building a more ideal implementation

e) ignore this problem and hope it goes away

Reply:
> I know of cases where falling back to a non-local endpoint is semantically wrong, and isn't acceptable. Setting a policy to "Local" ought to be a rule

Fair enough

b) seems the simplest solution